Add Support for TensorRT 8

MattConley opened the pull request:
This PR adds TRT 8 support to TensorFlow, removing deprecated API calls and updating TFTRT's SimpleTensor, FakeTensor, and ITensor handling with a new ITensorProxyPtr class. There will be a follow up PR to remove the use of ITensorProxyPtr class, see https://github.com/tensorflow/tensorflow/pull/48917#issuecomment-864244135.

Tagging @bixia1 @DEKHTIARJonathan @tfeher @jhalakpatel  for visibility

jhalakpatel commented:
@googlebot I fixed it.

MattConley committed:
Wrap TF-TRT Code in Proper Checks

gbaned commented:
@MattConley Can you please sign CLA. Thanks!

DEKHTIARJonathan commented:
@googlebot I fixed it.

jhalakpatel commented:
@googlebot I fixed it.

jhalakpatel commented:
@googlebot I consent.

DEKHTIARJonathan commented:
FYI @bixia1 CLA issue fixed ;)

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,599 @@
+    }
+  }
+}
+// A fake ITensor implementation used to check whether the TF-TRT converter can
+// handle specific node. We only need shape and type information, and the
+// converter won't (and shouldn't) use this to build the TRT network.
```
(1) Let's revise the comment:

SimpleITensor implements part of the ITensor interfaces to support the TF-TRT validator. As such, it only implements the interfaces that are related to shape and type information.

(2) can we remove the member functions that are not needed? For example, according to the comment, isNetworkInput shouldn't be needed. If that is not the case, we should fix the comment.

MattConley replied:
Since we merged in the former test class FakeTensor, I think it might be best to leave the definitions of methods unused by the converter in case they're relevant to any tests (current or future).  Given that, I've updated the comment to say: `SimpleITensor implements part of the ITensor interfaces to support the TF-TRT validator, as well as some TF-TRT tests. The former use case only utilizes the interfaces related to shape and type information.`

Does this seem like a reasonable approach?

bixia1 replied:
I am fine with this.

MattConley replied:
Sounds good, thanks!

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,599 @@
+  float dynamic_range_;
+};
+
+// Fake ITensor implementation for testing purposes.
+class FakeITensor {
```
This class is the same as SimpleITensor, is it? Maybe should remove this class and use SimpleITensor in the test?

MattConley replied:
Good point; merged the two classes in the latest commit, thanks

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,599 @@
+
+enum class TensorType : int
+{
+  kNONE,
```
We shouldn't need kNONE, right?

MattConley replied:
Shifting over to an exhaustive switch as you mention below sounds like an improvement, resolving this conversation and using that comment for the changes.

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,599 @@
+
+  TensorType ttype_{TensorType::kNONE};
+
+  // When ITensorProxy represents an ITensor, the ITensor can be either passed
+  // by the caller via the constructor that takes an ITensor* as parameter, or
+  // be created as a SimpleITensor.
+  //
+  // In the first case, the ITensor pointer is stored in 'tensor_' below, and
+  // the ITensor itself is not owned by this class. This method is used by
+  // Converter (e.g. AddInputTensor) and op converters during TRT network
+  // construction, where the TRT network owns the ITensor.
+  //
+  nvinfer1::ITensor* trt_tensor_ = nullptr; // Not owned.
+  // In the second case, the created SimpleITensor is stored in
+  // 'simple_itensor_' below and is owned by this class. SimpleITensor is a fake
+  // implementation of ITensor and is used only by TrtNodeValidator to validate
+  // the graph nodes.
+  std::shared_ptr<SimpleITensor> simple_tensor_ = nullptr;
+  std::unique_ptr<FakeITensor> fake_tensor_ = nullptr;
```
Shouldn't they be in a union?

MattConley replied:
This is done with the class merge, thanks

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,599 @@
+  std::unique_ptr<FakeITensor> fake_tensor_ = nullptr;
+};
+
+class ITensorProxyPtr
```
I think we don't need this class, instead, just use ITensorProxy which is already a wrapper of a pointer.

MattConley replied:
I like this idea as well, I'm putting together those changes for some tests

MattConley replied:
These changes are scheduled for the follow up PR mentioned here: https://github.com/tensorflow/tensorflow/pull/48917#issuecomment-864244135

bixia1 reviewed (commented):
Thanks for working on this! Would like to send out the first round of comment so that we can settle down on ITensorProxy before going deeper.

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,599 @@
+  // 'simple_itensor_' below and is owned by this class. SimpleITensor is a fake
+  // implementation of ITensor and is used only by TrtNodeValidator to validate
+  // the graph nodes.
+  std::shared_ptr<SimpleITensor> simple_tensor_ = nullptr;
```
I think we should change this to SimpleITensor* simple_tensor_, and let TRT_TensorOrWeights owns the object.

MattConley replied:
Now that the former FakeTensor class is now incorporated into SimpleTensor, would it be better to leave the SimpleTensor owned by ITensorProxy for any tests that don't involve the converter's use of TRT_TensorOrWeights?

bixia1 replied:
I think the answer depends on how you want to position ITensorProxy. To me, ITensorProxy is to provide a program interface for treating SimpleITensor and nvinfers::ITensor the same way (for TF-TRT validation and conversion). As it is a proxy, it should be copiable. If you agree on this, we should NOT make ITensorProxy owning the object to SimpleITensor. Instead, callers, such as the tests and TRT_TensorOrWeights should create and own SimpleITensor.

MattConley replied:
Makes sense; this change is scheduled to be added in the follow up PR mentioned here: https://github.com/tensorflow/tensorflow/pull/48917#issuecomment-864244135

bixia1 reviewed (approved):
Temporarily approving it for myself to get a copy of it. Will revert the approval once I have done this.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -849,87 +857,31 @@ string TRT_ShapedWeights::DebugString() const {
                                          const nvinfer1::Dims& trt_dims,
                                          int batch_size)
-    : simple_itensor_(new SimpleITensor(trt_dtype, trt_dims)),
+    : tensor_proxy_ptr_(new SimpleITensor(trt_dtype, trt_dims)),
```
Who is going to release the memory?
Similar in other places where you  use "new", we need to make sure that the memory will be released.

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,599 @@
+
+  bool isNetworkInput() const
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->isNetworkInput();
+      case TensorType::kSIMPLE: return simple_tensor_->isNetworkInput();
+      case TensorType::kFAKE: return fake_tensor_->isNetworkInput();
+      case TensorType::kNONE: assert(0 && "Unsupported itensor_ type");
+    }
+    assert(0 && "Should not come here.");
```
This routine can't be compiled, you need to return a value at the end as well. 
A much better way is to use exhaustive switch, see here https://abseil.io/tips/147
for the case here, you can add a "return" on the kNONE branch. 
Previously, I also suggest to remove kNONE from the enum as there is no such a need. I also suggest to remove kFAKE (only used by tests)and make kSIMPLE work for the tests.

Similar problems in many other routine in this file.

MattConley replied:
Thanks for the suggestion, after implementing this, removing the kNONE type, and merging the simple and fake tensors, these checks look much more streamlined

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,599 @@
+    
+  //! SimpleITensor owned
+  explicit ITensorProxy(nvinfer1::DataType trt_dtype, const nvinfer1::Dims& trt_dims)
+    : simple_tensor_(std::unique_ptr<SimpleITensor>(new SimpleITensor(trt_dtype, trt_dims))),
+      ttype_(TensorType::kSIMPLE) {}
```
This code can't be compiled. Please note that the order of the class member initialization needs to be the same as the order of the class member declaration. Similar in a few other constructors here.

MattConley replied:
Good catch, I've moved the tensor type to be the last declared member of the class which looks to take care of the issue

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,599 @@
+public:
+  ITensorProxyPtr(nullptr_t) : p_(nullptr) {}
+  ITensorProxyPtr(ITensorProxy* p) : p_(p) {}
+  ITensorProxyPtr(nvinfer1::ITensor* p) : p_(new ITensorProxy(p)) {}
+  ITensorProxyPtr(SimpleITensor* p) : p_(new ITensorProxy(p)) {}
+  
+  // FakeITensor
+  ITensorProxyPtr(FakeITensor* p) : p_(new ITensorProxy(p)) {}
+  ITensorProxyPtr() : p_(new ITensorProxy()) {}
+  ITensorProxyPtr(const nvinfer1::Dims& dims) : p_(new ITensorProxy(dims)) {}
+  ITensorProxyPtr(const std::vector<int>& dims) : p_(new ITensorProxy(dims)) {}
+
+  ITensorProxy* p_{nullptr};
```
Who is going to delete the storage allocated via new in this class? I see the object ITensorProxyPtr is copied, for example in convert_nodes.cc line 467 "Status DynamicBroadcast(ITensorProxyPtr operand"

MattConley replied:
Good point; I've switched the ITensorProxy reference to be a smart pointer instead, which should handle the new memory.

MattConley replied:
The original implementation of this fix was buggy, and so it is scheduled to be added in the follow up PR mentioned in this comment: https://github.com/tensorflow/tensorflow/pull/48917#issuecomment-864244135

MattConley commented:
Thanks for the reviews, I'm putting together an update to address the comments, and will address any of the other questions in the reviews themselves.

gbaned commented:
@MattConley  Can you please resolve conflicts? Thanks!

MattConley committed:
Merge SimpleTensor and FakeTensor implementations

- Combines both ITensorProxyPtr utility classes into one utilization

bixia1 commented on tensorflow/compiler/tf2tensorrt/tensorrt_test.cc:
```diff
@@ -22,29 +22,19 @@ limitations under the License.
 
-namespace tensorflow {
-namespace {
+//TODO(mconley): Should the below macro and the one in convert/utils.h be moved to an upper utils directory?
```
Good point, there is a similar macro in tf2tensorrt/utils/trt_logger.cc and one in trt_tensor_proxy.h.
We can put it to tf2tensorrt/common/utils.h.

MattConley replied:
Sounds good, I've moved the macro there and adjusted the imports accordingly

bixia1 reviewed (commented):
I see that there are un-addressed comments. Please do let me know when the PR is ready for another round of review. Please also check the tests with TRT7 and TRT8.

gbaned commented:
@MattConley  Can you please check @bixia1's comments and keep us posted ? Thanks!

MattConley committed:
Update ITensorProxy utility

- Remove kNONE tensor type in favor of exhaustive switch
- Fix order of parameter declaration

MattConley committed:
Move TRT Version Check Macro

MattConley commented:
Thanks for the help! I'll have one more set of pushes to address the remaining comments and then we should be ready for the next round of review.

gbaned commented:
@MattConley  Can you please resolve conflicts? Thanks!

MattConley committed:
Update SimpleTensor Description Comment

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,489 @@
+
+namespace tensorrt {
+
+namespace {
+  nvinfer1::Dims GetTestDims(const std::vector<int>& d) {
+    nvinfer1::Dims dims;
+    dims.nbDims = d.size();
+    for (int i = 0; i < d.size(); ++i) {
+      dims.d[i] = d[i];
+    }
+  }
+}
```
Using unnamed namespace in header can cause problem (see [here](https://rules.sonarsource.com/cpp/RSPEC-1000)) and is prohibited by c++ coding style ([here](https://google.github.io/styleguide/cppguide.html#Internal_Linkage)).

DEKHTIARJonathan commented:
## Daily Status Update - 06/15/2021

COMMIT ID TESTED: 59b148b6014

CC: @mattconley @bixia1 @tfeher @jhalakpatel

### Build Status:
- **TRT 7.1.3** - Current Google Default: :heavy_check_mark: PASS
- **TRT 7.2.3.4** - Current NVIDIA Default: :heavy_check_mark: PASS
- **TRT 8.x** - Upcoming release: :heavy_check_mark: PASS

### C++ Unittests Status:
- **TRT 7.1.3** - Current Google Default: :x: FAIL 8/16
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test_gpu
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test_gpu
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test_gpu
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test_gpu

- **TRT 7.2.3.4** - Current NVIDIA Default: :x: FAIL 8/16
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test_gpu
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test_gpu
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test_gpu
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test_gpu
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test

- **TRT 8.x** - Upcoming release: :x: FAIL 8/16
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test_gpu
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test_gpu
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test_gpu
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test_gpu
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test

### Python Unittests Status:
- **TRT 7.1.3** - Current Google Default: :x: FAIL
  - //tensorflow/python/compiler/tensorrt:trt_convert_test
  - //tensorflow/python/compiler/tensorrt:trt_convert_test_gpu
  
- **TRT 7.2.3.4** - Current NVIDIA Default: :x: FAIL
  - //tensorflow/python/compiler/tensorrt:trt_convert_test
  - //tensorflow/python/compiler/tensorrt:trt_convert_test_gpu
  
- **TRT 8.x** - Upcoming release: :x: FAIL
  - //tensorflow/python/compiler/tensorrt:trt_convert_test
  - //tensorflow/python/compiler/tensorrt:trt_convert_test_gpu

MattConley commented:
@bixia1 Thanks for your continued review, I'm addressing the comments now.
@DEKHTIARJonathan Thanks for your test; the shared pointer implementation I pushed earlier appears to have worsened the memory issues rather than fixing them, so I've reverted it for now and will push an updated version as soon as I confirm it passes the tests.

DEKHTIARJonathan commented:
## Daily Status Update - ##/##/2021

COMMIT ID TESTED: 22971b4eb07

CC: @mattconley @bixia1 @tfeher @jhalakpatel

### Build Status:
- **TRT 7.1.3** - Current Google Default: :heavy_check_mark: PASS
- **TRT 7.2.3.4** - Current NVIDIA Default: :heavy_check_mark: PASS
- **TRT 8.x** - Upcoming release: :heavy_check_mark: PASS

### C++ Unittests Status:
- **TRT 7.1.3** - Current Google Default: :heavy_check_mark: PASS 14/16 | :x: FAIL 02/16
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test_gpu                            :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test                                :x: FAIL
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test_gpu                            :x: FAIL
  - //tensorflow/compiler/tf2tensorrt:segment_test                                      :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:segment_test_gpu                                  :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:tensorrt_test_cc                                  :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:tensorrt_test_cc_gpu                              :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_allocator_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test_gpu                            :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test                      :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test_gpu                  :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_lru_cache_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_shape_optimization_profiles_test              :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_shape_optimization_profiles_test_gpu          :heavy_check_mark: PASS

- **TRT 7.2.3.4** - Current NVIDIA Default: :heavy_check_mark: PASS ##/16 | :x: FAIL ##/16
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test_gpu                            :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test                                :x: FAIL
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test_gpu                            :x: FAIL
  - //tensorflow/compiler/tf2tensorrt:segment_test                                      :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:segment_test_gpu                                  :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:tensorrt_test_cc                                  :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:tensorrt_test_cc_gpu                              :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_allocator_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test_gpu                            :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test                      :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test_gpu                  :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_lru_cache_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_shape_optimization_profiles_test              :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_shape_optimization_profiles_test_gpu          :heavy_check_mark: PASS


- **TRT 8.x** - Upcoming release: :heavy_check_mark: PASS ##/16 | :x: FAIL ##/16
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:convert_graph_test_gpu                            :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:convert_nodes_test_gpu                            :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:segment_test                                      :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:segment_test_gpu                                  :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:tensorrt_test_cc                                  :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:tensorrt_test_cc_gpu                              :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_allocator_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_op_test_gpu                            :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test                      :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test_gpu                  :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_lru_cache_test                                :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_shape_optimization_profiles_test              :heavy_check_mark: PASS
  - //tensorflow/compiler/tf2tensorrt:trt_shape_optimization_profiles_test_gpu          :heavy_check_mark: PASS


### Python Unittests Status:
- **TRT 7.1.3** - Current Google Default: :heavy_check_mark: PASS 02/02 | :x: FAIL 00/02
  - //tensorflow/python/compiler/tensorrt:trt_convert_test                              :heavy_check_mark: PASS
  - //tensorflow/python/compiler/tensorrt:trt_convert_test_gpu                          :heavy_check_mark: PASS
  
- **TRT 7.2.3.4** - Current NVIDIA Default: :heavy_check_mark: PASS ##/02 | :x: FAIL ##/02
  - //tensorflow/python/compiler/tensorrt:trt_convert_test                              :heavy_check_mark: PASS
  - //tensorflow/python/compiler/tensorrt:trt_convert_test_gpu                          :heavy_check_mark: PASS
  
- **TRT 8.x** - Upcoming release: :heavy_check_mark: PASS ##/02 | :x: FAIL ##/02
  - //tensorflow/python/compiler/tensorrt:trt_convert_test                              :heavy_check_mark: PASS
  - //tensorflow/python/compiler/tensorrt:trt_convert_test_gpu                          :heavy_check_mark: PASS
  
## Python Test Files Status:
- **TRT 7.1.3** - Current Google Default: :heavy_check_mark: PASS 23/27 | :x: FAIL 02/27 | :white_check_mark: IGNORED 02/27
  - /tensorflow/python/compiler/tensorrt/test/annotate_max_batch_sizes_test.py          :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/base_test.py                              :x: FAIL `Cannot register 2 metrics with the same name`
  - /tensorflow/python/compiler/tensorrt/test/batch_matmul_test.py                      :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/biasadd_matmul_test.py                    :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/binary_tensor_weight_broadcast_test.py    :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/cast_test.py                              :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/combined_nms_test.py                      :x: FAIL: `core dump => CombinedNmsTest.testTfTrtV2_OfflineConversion_DynamicEngine_FP16_NoCalibration`
  - /tensorflow/python/compiler/tensorrt/test/concatenation_test.py                     :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/const_broadcast_test.py                   :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/conv2d_test.py                            :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/dynamic_input_shapes_test.py              :white_check_mark: IGNORED
  - /tensorflow/python/compiler/tensorrt/test/identity_output_test.py                   :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/int32_test.py                             :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/lru_cache_test.py                         :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/memory_alignment_test.py                  :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/multi_connection_neighbor_engine_test.py  :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/neighboring_engine_test.py                :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/quantization_mnist_test.py                :white_check_mark: IGNORED
  - /tensorflow/python/compiler/tensorrt/test/quantization_test.py                      :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/rank_two_test.py                          :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/reshape_transpose_test.py                 :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/tf_function_test.py                       :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/topk_test.py                              :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/trt_mode_test.py                          :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/unary_test.py                             :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/vgg_block_nchw_test.py                    :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/vgg_block_test.py                         :heavy_check_mark: PASS
  
- **TRT 7.2.3.4** - Current NVIDIA Default: :heavy_check_mark: PASS ##/27 | :x: FAIL ##/27 | :white_check_mark: IGNORED ##/27
  - /tensorflow/python/compiler/tensorrt/test/annotate_max_batch_sizes_test.py          :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/base_test.py                              :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/batch_matmul_test.py                      :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/biasadd_matmul_test.py                    :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/binary_tensor_weight_broadcast_test.py    :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/cast_test.py                              :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/combined_nms_test.py                      :x: FAIL: `core dump => CombinedNmsTest.testTfTrtV2_OfflineConversion_DynamicEngine_FP16_NoCalibration`
  - /tensorflow/python/compiler/tensorrt/test/concatenation_test.py                     :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/const_broadcast_test.py                   :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/conv2d_test.py                            :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/dynamic_input_shapes_test.py              :white_check_mark: IGNORED
  - /tensorflow/python/compiler/tensorrt/test/identity_output_test.py                   :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/int32_test.py                             :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/lru_cache_test.py                         :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/memory_alignment_test.py                  :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/multi_connection_neighbor_engine_test.py  :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/neighboring_engine_test.py                :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/quantization_mnist_test.py                :white_check_mark: IGNORED
  - /tensorflow/python/compiler/tensorrt/test/quantization_test.py                      :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/rank_two_test.py                          :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/reshape_transpose_test.py                 :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/tf_function_test.py                       :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/topk_test.py                              :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/trt_mode_test.py                          :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/unary_test.py                             :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/vgg_block_nchw_test.py                    :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/vgg_block_test.py                         :heavy_check_mark: PASS
  
- **TRT 8.x** - Upcoming release: :heavy_check_mark: PASS 25/27 | :x: FAIL 00/27 | :white_check_mark: IGNORED 02/27
  - /tensorflow/python/compiler/tensorrt/test/annotate_max_batch_sizes_test.py          :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/base_test.py                              :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/batch_matmul_test.py                      :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/biasadd_matmul_test.py                    :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/binary_tensor_weight_broadcast_test.py    :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/cast_test.py                              :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/combined_nms_test.py                      :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/concatenation_test.py                     :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/const_broadcast_test.py                   :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/conv2d_test.py                            :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/dynamic_input_shapes_test.py              :white_check_mark: IGNORED
  - /tensorflow/python/compiler/tensorrt/test/identity_output_test.py                   :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/int32_test.py                             :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/lru_cache_test.py                         :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/memory_alignment_test.py                  :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/multi_connection_neighbor_engine_test.py  :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/neighboring_engine_test.py                :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/quantization_mnist_test.py                :white_check_mark: IGNORED
  - /tensorflow/python/compiler/tensorrt/test/quantization_test.py                      :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/rank_two_test.py                          :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/reshape_transpose_test.py                 :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/tf_function_test.py                       :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/topk_test.py                              :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/trt_mode_test.py                          :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/unary_test.py                             :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/vgg_block_nchw_test.py                    :heavy_check_mark: PASS
  - /tensorflow/python/compiler/tensorrt/test/vgg_block_test.py                         :heavy_check_mark: PASS

bixia1 reviewed (approved):
Temporarily approving it to get it to google conveniently. Will revert the approval right after get the code.

bixia1 reviewed (changes requested):
I couldn't get it to google for testing due to this https://github.com/tensorflow/tensorflow/pull/48917#issuecomment-864044532

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_allocator.h:
```diff
@@ -53,8 +53,8 @@ class TRTDeviceAllocator : public TRTBaseAllocator {
   }
-  void* allocate(uint64_t size, uint64_t alignment, uint32_t flags) override;
-  void free(void* memory) override;
+  void* allocate(uint64_t size, uint64_t alignment, uint32_t flags) noexcept override;
+  void free(void* memory) noexcept override;
```
You also add noexcept to the implementation, but I don't see any corresponding change to trt_allocator.cc.

MattConley replied:
The cc files mentioned all have matching noexcept tags now, thanks!

bixia1 reviewed (commented):
I couldn't get it to google for testing due to this https://github.com/tensorflow/tensorflow/pull/48917#issuecomment-864044532

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_logger.h:
```diff
@@ -28,7 +28,7 @@ namespace tensorrt {
  public:
   Logger(string name = "DefaultLogger") : name_(name) {}
-  void log(nvinfer1::ILogger::Severity severity, const char* msg) override;
+  void log(nvinfer1::ILogger::Severity severity, const char* msg) noexcept override;
```
Similar, need to add noexcept to the implementation of log in the cc file.

MattConley replied:
Thanks, added

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,482 @@
+namespace tensorrt {
+
+namespace {
+  nvinfer1::Dims GetTestDims(const std::vector<int>& d) {
+    nvinfer1::Dims dims;
+    dims.nbDims = d.size();
+    for (int i = 0; i < d.size(); ++i) {
+      dims.d[i] = d[i];
+    }
```
I am curious why this code work for you. It should return a value but doesn't.

MattConley replied:
Definitely would have expected this to fail as well.  It's possible that the method in convert_nodes_test.cc with the same signature (but not in the empty namespace) somehow conflated things?  It's fixed now, and the function is slightly renamed to 'ConvertVectorDims' as to avoid conflicts.

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,482 @@
+private:
+  bool validate() const
+  {
+    return trt_tensor_ && !simple_tensor_ || !trt_tensor_ && simple_tensor_;
```
Need to add parenthesis, ( trt_tensor_ && !simple_tensor_&&  ) || (!trt_tensor_ && simple_tensor_)  to avoid this warning https://stackoverflow.com/questions/16939888/why-does-clang-warn-within

MattConley replied:
Good point, added now

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_int8_calibrator.h:
```diff
@@ -69,9 +69,9 @@ struct TRTInt8Calibrator : public nvinfer1::IInt8EntropyCalibrator {
+  const void* readCalibrationCache(std::size_t& length) noexcept override;
 
-  void writeCalibrationCache(const void* ptr, std::size_t length) override;
+  void writeCalibrationCache(const void* ptr, std::size_t length) noexcept override;
```
please also add noexcept to the implementation.

MattConley replied:
Added

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -849,87 +857,31 @@ string TRT_ShapedWeights::DebugString() const {
-  nvinfer1::Dims trt_dims_;
-};
+TRT_TensorOrWeights::TRT_TensorOrWeights(ITensorProxyPtr tensor, int batch_size)
+    : tensor_proxy_ptr_(tensor),
+      batch_size_(batch_size),
```
In the class definition, batch_size_ comes before tensor_proxy_ptr_, you need to initialize the fields in the same order they are declared. Similar problems in a few constructors below.

MattConley replied:
Thanks for the recommendation, moved the tensor_proxy_ptr_ as suggested, and the rest of TRT_TensorOrWeights's members are in order

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.h:
```diff
@@ -615,10 +604,13 @@ class Converter {
                     absl::string_view sub_op_name,
-                    absl::optional<int> sub_op_instance = absl::nullopt);
+                    absl::optional<int> sub_op_instance = absl::nullopt,
+                    absl::optional<std::string> origin_node_name = absl::nullopt);
```
I think this parameter is NOT added to the implementation, see the cc file. I couldn't comment in the implementation because it is not changed by this PR.

MattConley replied:
Ah this was incorrectly applied to the second SetLayerName function instead of PrepareTensorForShape.  Fixed now, thanks

MattConley commented:
@googlebot I fixed it

jhalakpatel commented:
@googlebot I fixed it.

jhalakp-nvidia commented:
@googlebot I fixed it.

MattConley commented:
This PR is the initial implementation of adding TensorRT 8 support to TensorFlow and TF-TRT.  After some discussion about the design of the newly added ITensorProxy and ITensorProxyPtr classes, we are planning on deferring some changes of this system to a second PR.  While the current implementation is functional, it does complicate the memory management of nvinfer1::ITensor and SimpleITensor objects, resulting in a less optimal and somewhat overly complex configuration.  We plan to remove the ITensorProxyPtr class in the follow up PR, simplifying the interactions to focus on the ITensorProxy class and using pointers thereto when necessary.  This change will also allow us to more easily redistribute ownership properly while avoiding memory leaks and an overabundance of shared and unique pointers.  One specific example mentioned earlier in the review of this PR which we will be addressing is moving the SimpleITensor object currently owned by ITensorProxy to TRT_TensorOrWeights.

MattConley commented:
@googlebot I fixed it.

MattConley committed:
Add Support for TRT 8

Co-authored-by: Jhalak Patel <jhalakpatel@outlook.com>

MattConley committed:
Minor Test Updates

MattConley committed:
Fix Errors and Address Reviewer Comments

MattConley commented:
@googlebot I fixed it.

MattConley commented:
@bixia1 Thanks for your review!  Please let me know if there's any additional information I can provide / changes I should make to the code.  The CLA issue was due to me accidentally using a different email for some of the more recent commits, but after rebasing to amend the author (and also removing the reverted smart pointer commit) it looks good now.

MattConley commented:
Working on an issue that just came up, fix incoming momentarily.

MattConley committed:
Test Fixes

MattConley commented:
Should be good to go now, thanks again

bixia1 reviewed (approved):
Temporarily approving it so that I can get the code to google.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -109,7 +109,10 @@ const char* LayerTypeToString(nvinfer1::LayerType layer_type) {
+#endif
     ADD_LAYER(UNARY)
     ADD_LAYER(PADDING)
     ADD_LAYER(SHUFFLE)
```
Need to add this to the end of the list:
#if IS_TRT_VERSION_GE(8, 0, 0, 0)
    ADD_LAYER(QUANTIZE)
    ADD_LAYER(DEQUANTIZE)
#endif

bixia1 replied:
I see code change for this. In the future, please response to the comment so I know you already addressed it.

MattConley replied:
Added the quantize conversions and moved the TRT8 check to the bottom to keep things slightly cleaner, thanks

bixia1 commented:
I see convert_nodes_test has a correctness issue and fails:
Expected equality of these values:
  input->getDynamicRangeMax()
    Which is: 0
  5.0f
    Which is: 5

There are a few other test with memory leak, see log using TRT8. Similar memory leak happens with TRT 7 also. See log.

bixia1 commented:
[fail4.log](https://github.com/tensorflow/tensorflow/files/6688593/fail4.log)
[fail3.log](https://github.com/tensorflow/tensorflow/files/6688594/fail3.log)
[fail2.log](https://github.com/tensorflow/tensorflow/files/6688596/fail2.log)
[fail1.log](https://github.com/tensorflow/tensorflow/files/6688599/fail1.log)

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:
```diff
@@ -0,0 +1,478 @@
+class ITensorProxyPtr
+{
+public:
+  ITensorProxyPtr(nullptr_t) : p_(nullptr) {}
+  ITensorProxyPtr(ITensorProxy* p) : p_(p) {}
+  ITensorProxyPtr(nvinfer1::ITensor* p) : p_(new ITensorProxy(p)) {}
+  ITensorProxyPtr(SimpleITensor* p) : p_(new ITensorProxy(p)) {}
+  
+  ITensorProxyPtr() : p_(new ITensorProxy()) {}
+  ITensorProxyPtr(const nvinfer1::Dims& dims) : p_(new ITensorProxy(dims)) {}
+  ITensorProxyPtr(const std::vector<int>& dims) : p_(new ITensorProxy(dims)) {}
+
+  ITensorProxy* p_{nullptr};
```
The new op here causes memory leak. you can fix it by make p_ std::shared_ptr, though this is no an idea solution. but i expect you to remove this class in the follow up PR.

MattConley replied:
We'll be removing the class in the follow up PR, but the shared_ptr has been added for now, thanks

MattConley committed:
Address Reviewer Comments

bixia1 reviewed (commented):
Can you also resolve conflict?

MattConley committed:
Merge remote-tracking branch 'upstream/master' into trt8_updates

MattConley commented:
Thanks for the continued review.  I've fixed the merge conflict and addressed the above reviews, but am still trying to reproduce and fix the DynamicRange issue you mention above.  I should have the changes for that pushed within the next few hours

MattConley committed:
Fix Tensor Proxy Dynamic Ranges

MattConley committed:
Test Updates

bixia1 reviewed (approved):
Approving this for myself to get it to google for testing.
